src/chainrules.jl

```diff
@@ -0,0 +1,158 @@
+## Forward Rules
+
+function ChainRulesCore.frule((_, Δx, Δy), d::Distances.Euclidean, x, y)
+    Δ = x - y
+    D = sqrt(sum(abs2, Δ))
+    if iszero(D)
+        return D, Zero()
+    else
+        Δ ./= D
+        return D, dot(Δ, Δx) - dot(Δ, Δy)
+    end
```

Comment:
This is definitely type unstable, isn't it? Maybe we could write instead:
```suggestion
    if !iszero(D)
        Δ ./= D
    end
    return D, dot(Δ, Δx) - dot(Δ, Δy)
```

Reply:
Ah yeah that's a good idea. But anyway it's not used right now.
Both ForwardDiff and ReverseDiff return `NaN` for the `Euclidean` metric, which is correct but highly undesirable. I can fix it for ReverseDiff but any idea how I can do it for ForwardDiff?